### PR TITLE
Simplify block behavior and improve account block feedback.

### PR DESCRIPTION
**Pull request by crhallberg**
This replaces PR #591 in resolving issue VUFIND-1145.

TODO
- [x] Send actual block reason to user when possible.
- [x] Consider whether adding a new public ILS driver method to get block messages would allow us to avoid changing the format returned by the holdings formatter and might also allow us to completely eliminate the special 'block' return value without losing functionality.
- [x] Ensure consistent behavior in jquerymobile theme.
- [x] Update driver spec to reflect new method and removal of addLink == block special case.

### CODE REVIEW

**Comment by demiankatz** [comment]
@crhallberg, I talked to @EreMaijala about this today. He likes the refactoring but suggested that it would be more useful if we could send the actual block reason out to the interface instead of using the generic message (when supported). Let's get together next week and see how easily we can extend the ILS driver interface to support this feature.

Also, I seem to recall that we're always putting the message into the template but applying a hidden class when it is not needed. Since we are no longer un-hiding it with Javascript, this is unnecessary. We should move the conditional around the entire message instead of applying the hidden class.

**Comment by crhallberg** [comment]
I think it may be as easy as setting the flag to the message instead of true where relevant. We'll see on Tuesday!

**Comment by demiankatz** [comment]
@EreMaijala, we've made some further progress on this. We've added a new optional ILS driver method called getRequestBlocks, which returns an array of block messages that prevent users from placing requests. All of the special-case 'block' logic has been removed from the ILS drivers, as this was confusing and redundant. Instead, we use getRequestBlocks in the hold logic to block (or not block) all request links. This also allows us to pass specific messages to the template. What do you think?

**Comment by EreMaijala** [comment]
Looks good to me, but I'd like to also have a getBlocks method for retrieving all the patron blocks regardless of whether they block requesting. Those could then be displayed in at least the loans and profile pages.
Having a separate getRequestBlocks is good since borrowing blocks might not prevent requesting.

**Comment by demiankatz** [comment]
I agree that having a more generic mechanism to get all blocks is a good idea. I just don't want to design that part of the ILS interface too hastily, because I'm not sure what granularity differences may exist between systems. Looking at the way blocks are retrieved in different ILS driver for the purposes of renewals suggests that this is definitely not uniform functionality. I suppose it would probably be safe to add a getAllBlocks() method as part of the interface, since that can always be expanded on a driver-by-driver basis... but I also wonder whether it would be better in the long term to define some constants and have a single getBlocks() method that could be filtered... but obviously this filtering idea requires us to figure out a generic cross-platform mechanism, while creating different methods for different types of blocks on an as-needed basis allows us to proceed more cautiously (at the cost of a more sprawling interface).

Bottom line: I guess we could proceed two different ways:

1.) Keep this as-is, but add getAllBlocks

2.) Change getRequestBlocks to getBlocks($filter = BLOCKS_ALL) and define BLOCKS_ALL and BLOCKS_REQUEST filters. We could use bitwise constants and make this work similar to set_error_handler, etc. Perhaps we could also add BLOCKS_RENEWAL and refactor all of the existing ILS drivers accordingly... though that makes me a little nervous without a reliable ability to test things.

What do you think?

**Comment by EreMaijala** [comment]
I'd like option 2, but since it's way more complex and there are risks involved, I'd go for option 1. Since it would be for display purposes, it doesn't need to be more complex.

**Comment by demiankatz** [comment]
Sounds good to me. We can let this grow organically for a while and then refactor it for a future major version release if we reach the point of feeling that it's safe to consolidate to a single function.

**Comment by demiankatz** [comment]
@EreMaijala, is https://github.com/vufind-org/vufind/pull/815/commits/104c8906dae6dab69d2905600146dcdf22697763 more or less what you had in mind?

**Comment by EreMaijala** [comment]
Yes, thanks!
